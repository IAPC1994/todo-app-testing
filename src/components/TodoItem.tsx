import { Todo } from "@/interfaces";
import clsx from "clsx";
import { IoTrashOutline } from "react-icons/io5";

interface Props {
  todo: Todo;
  onToggleTodo: (id: string) => void;
  onDeleteTodo: (id: string) => void;
}

export const TodoItem = ({ todo, onToggleTodo, onDeleteTodo }: Props) => {
  const { id, title, completed } = todo;
  return (
    <div className="flex items-center gap-2 bg-indigo-700 p-4 shadow-md rounded-lg">
      <input
        type="checkbox"
        className="w-5 h-5 md:w-8 md:h-8 accent-emerald-600 cursor-pointer"
        checked={completed}
        onChange={() => onToggleTodo(id)}
      />
      <p
        className={clsx("text-base md:text-xl flex-1", {
          "line-through text-gray-400": completed,
        })}
      >
        {title}
      </p>
      <button
        type="button"
        onClick={() => onDeleteTodo(id)}
        className="p-2 border rounded-lg shadow-md bg-red-500 hover:bg-red-600 cursor-pointer"
      >
        <IoTrashOutline size={20} />
      </button>
    </div>
  );
};
