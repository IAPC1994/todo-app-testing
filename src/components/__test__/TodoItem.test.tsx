import { fireEvent, render } from "@testing-library/react";
import { TodoItem } from "../TodoItem";

const mockTodo = {
  id: "1",
  title: "Todo 1",
  completed: false,
};

describe("Testing in <TodoItem /> component", () => {
  test("should be equal to snapshot", () => {
    const item = render(
      <TodoItem
        todo={mockTodo}
        onDeleteTodo={jest.fn()}
        onToggleTodo={jest.fn()}
      />
    );

    expect(item).toMatchSnapshot();
  });

  test("should render the Todo title passed via props", () => {
    const { getByText } = render(
      <TodoItem
        todo={mockTodo}
        onDeleteTodo={jest.fn()}
        onToggleTodo={jest.fn()}
      />
    );
    expect(getByText("Todo 1")).toBeInTheDocument();
  });

  test("should apply completed styles to the title when Todo is marked as complete", () => {
    const onToggleTodo = jest.fn();

    const { getByRole, getByText, rerender } = render(
      <TodoItem
        todo={mockTodo}
        onToggleTodo={onToggleTodo}
        onDeleteTodo={jest.fn()}
      />
    );
    const checkbox = getByRole("checkbox");
    const titleParagraph = getByText("Todo 1");
    expect(titleParagraph.className).not.toContain("line-through");
    expect(titleParagraph.className).not.toContain("text-gray-400");

    fireEvent.click(checkbox);

    expect(onToggleTodo).toHaveBeenCalledWith("1");

    rerender(
      <TodoItem
        todo={{ ...mockTodo, completed: true }}
        onToggleTodo={onToggleTodo}
        onDeleteTodo={jest.fn()}
      />
    );

    const updatedTitleParagraph = getByText("Todo 1");

    expect(updatedTitleParagraph.className).toContain("line-through");
    expect(updatedTitleParagraph.className).toContain("text-gray-400");
  });

  test("should called the onDeleteTodo and delete the Todo of the document", () => {
    const onDeleteTodo = jest.fn();
    const { getByRole, rerender, queryByText } = render(
      <TodoItem
        todo={mockTodo}
        onToggleTodo={jest.fn()}
        onDeleteTodo={onDeleteTodo}
      />
    );
    const deleteButton = getByRole("button");
    fireEvent.click(deleteButton);
    expect(onDeleteTodo).toHaveBeenCalledWith("1");

    rerender(<></>);
    expect(queryByText("Todo 1")).not.toBeInTheDocument();
  });
});
