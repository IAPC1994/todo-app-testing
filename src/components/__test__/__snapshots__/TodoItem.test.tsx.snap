// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <TodoItem /> component should be equal to snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex items-center gap-2 bg-indigo-700 p-4 shadow-md rounded-lg"
      >
        <input
          class="w-5 h-5 md:w-8 md:h-8 accent-emerald-600 cursor-pointer"
          type="checkbox"
        />
        <p
          class="text-base md:text-xl flex-1"
        >
          Todo 1
        </p>
        <button
          class="p-2 border rounded-lg shadow-md bg-red-500 hover:bg-red-600 cursor-pointer"
          data-testid="delete-btn-1"
          type="button"
        >
          <svg
            fill="currentColor"
            height="20"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m112 112 20 320c.95 18.49 14.4 32 32 32h184c17.67 0 30.87-13.51 32-32l20-320"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
            <path
              d="M80 112h352"
              stroke-linecap="round"
              stroke-miterlimit="10"
              stroke-width="32"
            />
            <path
              d="M192 112V72h0a23.93 23.93 0 0 1 24-24h80a23.93 23.93 0 0 1 24 24h0v40m-64 64v224m-72-224 8 224m136-224-8 224"
              fill="none"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="32"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex items-center gap-2 bg-indigo-700 p-4 shadow-md rounded-lg"
    >
      <input
        class="w-5 h-5 md:w-8 md:h-8 accent-emerald-600 cursor-pointer"
        type="checkbox"
      />
      <p
        class="text-base md:text-xl flex-1"
      >
        Todo 1
      </p>
      <button
        class="p-2 border rounded-lg shadow-md bg-red-500 hover:bg-red-600 cursor-pointer"
        data-testid="delete-btn-1"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m112 112 20 320c.95 18.49 14.4 32 32 32h184c17.67 0 30.87-13.51 32-32l20-320"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
          <path
            d="M80 112h352"
            stroke-linecap="round"
            stroke-miterlimit="10"
            stroke-width="32"
          />
          <path
            d="M192 112V72h0a23.93 23.93 0 0 1 24-24h80a23.93 23.93 0 0 1 24 24h0v40m-64 64v224m-72-224 8 224m136-224-8 224"
            fill="none"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="32"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
